cluster/saltbase/salt/kube-addons/kube-addon-update.sh

```diff
@@ -0,0 +1,390 @@
+            if [[ "${objBasename}-${new_suffix}" == "${name_from_file}" ]]; then
+                log DB3 "matched existing replication controller $obj_on_server to file $obj; suffix=$suffix, new_suffix=$new_suffix"
+                if [[ "$suffix" == "$new_suffix" ]]; then
+                    for_ignore="$for_ignore $name_from_file"
+                    matched_files="$matched_files $obj"
+                    found=1
+                    break
+                else
+                    for_update="$for_update $obj_on_server;$obj"
+                    matched_files="$matched_files $obj"
+                    found=1
+                    break
+                fi
+            fi
+        done
+        if [[ $found  == 0 ]]; then
+            log DB2 "No definition file found for replication controller $obj_on_server. Scheduling for deletion"
+            for_delete="$for_delete $obj_on_server"
+        fi
+    done
+
+
+    for obj in  $in_files; do
+        echo $matched_files | grep $obj >/dev/null
+        if [[ $? -ne 0 ]]; then
+            new_files="$new_files $obj"
+        fi
+    done
+}
+
+
+function reconcile-objects() {
+    local -r addon_path=$1
+    local -r obj_type=$2
+    match-objects $addon_path $obj_type
+
+    log DBG "$obj_type: for_delete=$for_delete"
+    log DBG "$obj_type: for_update=$for_update"
+    log DBG "$obj_type: for_ignore=$for_ignore"
+    log DBG "$obj_type: new_files=$new_files"
+
+    stop-objects "$obj_type" "$for_delete"
+    # wait for jobs below is a protection against changing the basename
+    # of a replication controllerm without changing the selector.
+    # If we don't wait, the new rc may be created before the old one is deleted
+    # In such case the old one will wait for all its pods to be gone, but the pods
+    # are created by the new replication controller. 
+    # passing --cascade=false could solve the problem, but we want 
+    # all orphan pods to be deleted.
+    wait-for-jobs
+    stopResult=$?
+
+    create-objects "$obj_type" "$new_files"
+    update-objects "$obj_type" "$for_update"
+
+    for obj in $for_ignore; do
+        log DB2 "The $obj_type $obj is already up to date"
+    done
+
+    wait-for-jobs
+    createUpdateResult=$?
+
+    if [[ $stopResult == 0 ]] && [[ $createUpdateResult == 0 ]]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+function update-addons() {
+    local -r addon_path=$1
+    # be careful, reconcile-objects uses global variables
+    reconcile-objects $addon_path ReplicationController &
+
+    # Warning, we don't expect service names to be versioned, so
+    # the update for services should be turned off - just new services
+    # should be created and the old ones should be gone
+    # Also, if service description differs, the service should be recreated
+    # regardless of its name. This is not implemented in this version.
+    reconcile-objects $addon_path Service &
```

Comment:
I'm not sure I understand the question.

for rc, we have, for instance, monitoring-heapster-v1 (basename is monitoring-heapster), so update will be performed to monitoring-heapster-v2 (rolling update, in the future).
For services we can't do it because we have monitoring-influxdb (and no version suffix).

We can:
1) rename the services to follow this convention (But services are bound in a lot of places - a lot of changes)
or
2) declare that there will be no update for services - just create a new service and delete the old one if the name changes (this makes sense because there will be no rolling update anyway)

I decided to go for 2).

I think that the same can be done for rc. the version suffix is not necessary. But then we have to make sure rolling update works also if rc name doesn't change.

If I had more time I'd remove the version suffix from replication controllers too. (maybe it's still worth doing for 1.0?)